OpenFGA authorization driver

markylaing opened the pull request:
This PR adds an OpenFGA authentication driver to LXD https://warthogs.atlassian.net/browse/LXD-389?atlOrigin=eyJpIjoiZWExM2NhNjgwMWU2NDZhMGJhODc5MjAzYmZiZGFkNTYiLCJwIjoiaiJ9. 

The spec can be found at https://discourse.ubuntu.com/t/openfga-authorization-driver/38979

Work still to be done:
- [x] Setting authentication model ID in database. Avoid race condition.
- [x] Permission sync remove dangling permissions.
- [x] Remove all usages of `context.Background()`.
- [x] Unit and system tests (also system tests for RBAC).
- [x] Documentation

tomponline commented:
@markylaing i know this is heavily under development for now, so no need to do this right away, but when it comes to review and merge, I would appreciate it if you could split this into two PRs; 1 for the interface and hook changes across the code base, and 2 introducing the new openfga authorization driver.

markylaing commented:
RBAC tests passing also :)

markylaing commented on lxd/auth/driver_openfga.go:
```diff
@@ -0,0 +1,933 @@
+
+	var builtinAuthorizationModel client.ClientWriteAuthorizationModelRequest
+
+	err = json.Unmarshal([]byte(authModel), &builtinAuthorizationModel)
```
### Migrations
Here `authModel` is the json string that is generated via `make update-openfga`. If there is an existing model in the store, we ensure that this model is exactly equal to the model we have statically defined. This is where we would have an opportunity to detect different model versions and apply update logic.

Migrations in OpenFGA follow a pretty basic idea. The idea is to create an intermediary model which supports both the new and the old, then write all the tuples required to support the new schema, delete all the tuples that are no longer required, and finally write the updated model (see https://openfga.dev/docs/modeling/migrating/migrating-relations).

In practice, I think we will likely have something like:
* Under an `lxd/auth/openfga` directory:
    - `model_v1.openfga`
    - `model_v1.go`
    - `model_v2_compatible.openfga`
    - `model_v2_compatible.go`
    - `model_v2.openfga`
    - `model_v2.go`
    - `migrations.go`
    - `synchronise.go`
* The `migrations.go` file will contain a map of methods which accept an openfga client and performs the necessary actions to migrate from one schema to the next. These methods will only be valid while the compatible model is in place.
* The `synchronise.go` file will contain a map of methods which accept an fga client and a set of LXD resources and ensure they match 1:1 with what is present in the OpenFGA store. Like migrations, this is also model specific.
* When we connect to the OpenFGA server:
    1. Get the current auth model.
    2. Check if it is the latest version. If it is, sync resources and continue. Otherwise....
    3. Write the compatible model.
    4. Call the associated method from `migrations.go`.
    5. Write the latest model.
    6. Perform a final resource sync from `synchronise.go`.

Hopefully we won't have to do this very often.

markylaing replied:
I think we can handle this when we come to update the model.

markylaing commented on lxd/auth/driver_openfga_model.openfga:
```diff
@@ -0,0 +1,108 @@
+  relations
+    define project: [project]
+    define can_edit: [user, group#member] or operator from project
+    define can_view: [user, group#member] or can_edit or viewer from project
```
For now viewing or editing warnings requires the `can_edit` requirement on `server` which is the highest level of permission.

markylaing commented on lxd/auth/driver_openfga_model.openfga:
```diff
@@ -0,0 +1,108 @@
+  relations
+    define project: [project]
+    define can_edit: [user, group#member] or operator from project
+    define can_view: [user, group#member] or can_edit or viewer from project
```
For now to list operations a user will need `can_view` on the project. Individual operations remain the same as they require a secret anyway.

markylaing commented:
Tests will fail until #12349 and #12347 are merged.

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+// ObjectByType returns an object of the given type. The passed in arguments must be in the correct
+// order (as found in the URL for the resource). This function will error if an invalid object type is
+// given, or if the correct number of arguments is not passed in. Project objects use the "projectName" argument.
+func ObjectByType(objectType ObjectType, projectName string, args ...string) (string, error) {
```
Could you add an example here for a valid "args" slice in the docblock here?

markylaing replied:
This has been removed.

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+	return "", fmt.Errorf("Unknown object type %q", objectType)
+}
+
+// ObjectFromRequest returns an object created from the request by evaluating the given mux vars.
```
The word "object" seems a bit misleading as this function returns a string. I realise Go doesn't have "objects" as such but often people refer to instantiated structs as objects. What is the object that this function returns? Can we clarify it by changing the type to a dedicated string type perhaps?

markylaing replied:
I'll try to change things a bit to make things a bit more obvious. I'll put some examples in here first to get consensus though. If only because rebasing this thing is starting to get tricky :laughing:

markylaing replied:
@tomponline I've added a WIP commit with some changes to this file to try and make things clearer. Can you please take a look at this and tell me what you think? If it looks good I'll make the change and rebase through both PRs. https://github.com/markylaing/lxd/blob/de51714968a469599a3ea272cf9433b2ccc57893/lxd/auth/authorization_objects.go

markylaing replied:
You can also view the WIP commit to see how this change affects other parts of the codebase https://github.com/canonical/lxd/pull/12252/commits/de51714968a469599a3ea272cf9433b2ccc57893

tomponline replied:
Yeah this looks good. I think it would be worth adding a comment to the `type Object string` definition explaining what it is though.

markylaing replied:
Yep I will do. The WIP commit is failing CI at the minute so on Monday morning I'll fix those errors, fix up comments etc. and have a look at the rest of your comments 🙂

tomponline replied:
Ta, I think the rest of the commits looked somewhat mechanical so ill wait until those comments have been addressed before reviewing rest of it.

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+// requires a project name, this is taken from the project query parameter unless the URL begins
+// with /1.0/projects.
+func ObjectFromRequest(r *http.Request, objectType ObjectType, muxVars ...string) (string, error) {
+	muxValues := make([]string, len(muxVars))
```
Can we use `make([]string,0, len(muxVars))` and then append instead of pre-allocating zero entries?

markylaing replied:
Done.

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+func ObjectFromRequest(r *http.Request, objectType ObjectType, muxVars ...string) (string, error) {
+	muxValues := make([]string, len(muxVars))
+	for i, muxVar := range muxVars {
+		muxValue, err := url.PathUnescape(mux.Vars(r)[muxVar])
```
Can you extract `mux.Vars(r)` once above rather than extracting it for every muxVars item?

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+	for i, muxVar := range muxVars {
+		muxValue, err := url.PathUnescape(mux.Vars(r)[muxVar])
+		if err != nil {
+			return "", err
```
Lets wrap this error to include the muxVar that was at fault to aid debugging issue when users report them.

markylaing replied:
Done.

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	// If using projects API we want to pass in the mux var, not the query parameter.
+	if objectType == ObjectTypeProject && strings.HasPrefix(r.URL.Path, "/1.0/projects") {
```
Can you use `shared.APIVersion` here instead of `1.0`

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+
+// ObjectParts deconstructs an object into an object type and a string slice of components that comprise the object.
+// An error is returned if the object does not have the expected format.
+func ObjectParts(object string) (ObjectType, []string, error) {
```
Similar to comment above, im still not really clear from reading this commit what an "object" is in this context, so I feel the comments and/or the type could be clarified here.

markylaing replied:
This has been removed.

tomponline commented on lxd/auth/authorization_objects.go:
```diff
@@ -0,0 +1,208 @@
+	}
+
+	escapedParts := strings.SplitN(objectName, "/", -1)
+	parts := make([]string, len(escapedParts))
```
Can you use `make([]string, 0,len(escapedParts))` and `append` instead of pre-allocating zero value slice entries?

markylaing replied:
Done.

tomponline commented on lxd/auth/authorization.go:
```diff
@@ -23,20 +32,61 @@ var authorizers = map[string]func() authorizer{
 
-	init(name string, config map[string]any, logger logger.Logger, projectsGetFunc func() (map[int64]string, error))
-	load() error
+	init(driver string, logger logger.Logger)
```
Can we call this `driverName` to align with the likes of `storage.Load` please.

tomponline commented on lxd/auth/authorization.go:
```diff
@@ -45,16 +95,38 @@ type UserAccess struct {
-	driverFunc, ok := authorizers[name]
+type Opts struct {
+	config          map[string]any
+	projectsGetFunc func(ctx context.Context) (map[int64]string, error)
```
Could we have a comment on this as to what its for, im not seeing why each driver needs its own project function?

markylaing replied:
I've added a comment above the `WithProjectsGetFunc` option function.

tomponline commented on lxd/auth/authorization.go:
```diff
@@ -45,16 +95,38 @@ type UserAccess struct {
+	}
+}
+
+func LoadAuthorizer(ctx context.Context, driver string, logger logger.Logger, options ...func(opts *Opts)) (Authorizer, error) {
```
Do we need to support different combinations of options via the variadic or can it be just static arguments?

tomponline commented on lxd/auth/driver_common.go:
```diff
@@ -5,15 +5,15 @@ import (
-	c.config = config
+func (c *commonAuthorizer) init(driver string, l logger.Logger) {
+	c.driver = driver
 	c.logger = l
```
Shall we wrap the logger with the driver name, ala

```
func (n *common) init() {
n.logger = logger.AddContext(logger.Ctx{"project": projectName, "driver": netInfo.Type, "network": netInfo.Name})
}
```

markylaing replied:
Done.

tomponline commented on lxd/auth/driver_tls.go:
```diff
@@ -1,65 +1,427 @@
+	}, nil
+}
+
+func (a *tls) relationToPermission(object string, entitlement Entitlement) (Permission, string, bool, error) {
```
In this function, is there a way to make it less maintenance heavy in the future?
What I mean is that TLS auth is pretty coarse, so do we need to define a mapping 1:1 between every object type and entitlement or can we make this more general so it doesn't need to be updated for every new object and entitlement?

markylaing replied:
I don't think it's going to be possible to generalise this unfortunately. The good news (as we discussed earlier) is that if we can't match the object and entitlement then this function will return an error, which will result in a `403 Forbidden`. We should be able to catch this easily in tests and if one slips by it'll be an easy fix.

markylaing replied:
I'm not sure what you mean. `relationToPermission` does extract the project from the object. This is used in the `UserHasPermission` method along with the permission to check the `UserAccess` struct for the permission on that project. The `UserAccess` struct is set in the request context using the users certificate before any authorization calls are made.

tomponline commented on lxd/auth/driver_rbac.go:
```diff
@@ -117,7 +118,7 @@ func (r *rbac) load() error {
 	go func() {
 		for {
-			err = r.syncProjects()
+			err = r.syncProjects(r.ctx)
```
We have a StopService() function, I wonder if we should also have a StartService() function, as Go recommends against having long-lived contexts inside a struct like the `load()` function suggests it does (I realise it doesn't actually do that though).

Also, having a StartService (that performs initial sync/activation) could return an error and avoid potentially LXD starting up in an inconsistent state with the authorizer service.

tomponline replied:
rather than `load` accepting a context.

tomponline replied:
I can see its retrying so that is nice too. Lets leave it as it is for now, but I think we should log an error at least if the sync process is failing.

tomponline replied:
It does beg the question, if we've not successfully synced yet, should we be fulfilling requests using this authorizer when the data its using for the check is potentially out of date? What do you think, should an authorizer have a "in sync" status that affects how it responds?

markylaing replied:
For the RBAC driver, the `UserAccess` it retrieved when a user authenticates. This forces a sync with the RBAC server and errors if the user is still not present in the local cache afterwards. The RBAC server also tells us if there has been a change on its end so we can resync everything.

Ultimately I didn't want to change the RBAC driver too much from what is existing for fear of breaking it.

tomponline commented on lxd/daemon.go:
```diff
@@ -260,6 +261,48 @@ func allowPermission(objectType auth.ObjectType, entitlement auth.Entitlement, m
+	return func(d *Daemon, r *http.Request) response.Response {
+		// If we don't have the appropriate driver, return response.EmptySyncResponse.
+		if d.authorizer.Driver() != auth.DriverTLS && d.authorizer.Driver() != auth.DriverRBAC {
+			return response.EmptySyncResponse
```
Does response.EmptySyncResponse mean "allow access"?

tomponline commented on lxd/daemon.go:
```diff
@@ -260,6 +261,48 @@ func allowPermission(objectType auth.ObjectType, entitlement auth.Entitlement, m
 
+// rbacPermission must be used as an access handler only when permissions that apply to RBAC/TLS drivers do not apply to
+// other drivers. If another driver is loaded, this will always return response.EmptySyncResponse.
+func rbacPermission(permission auth.Permission) func(d *Daemon, r *http.Request) response.Response {
```
@markylaing could you explain the differences that you couldn't resolve between rbac and openfga please? Thanks

markylaing replied:
This is to maintain the behaviour of TLS and RBAC drivers when we list API resources. On say `GET /1.0/instances` the TLS and RBAC drivers will return `403 Forbidden` if the user does not have `view` permission on the project (and if all-projects is used, the user must be an admin). Whereas with OpenFGA we will filter out all instances that the user cannot view, likely returning an empty list.

markylaing replied:
https://github.com/canonical/lxd/pull/12313#discussion_r1336893508

tomponline commented on lxd/db/cluster/entities.go:
```diff
@@ -6,24 +6,24 @@ import (
-	TypeStorageVolumeBackup   = 14
-	TypeStorageVolumeSnapshot = 15
-	TypeWarning               = 16
-	TypeClusterGroup          = 17
```
I actually prefer the explicit nature of the previous version because it means that if for any reason we remove a type in the future it wont reshuffle the IDs implicitly.

markylaing replied:
I didn't think the actual IDs would be important since we should always be using the constants but I have dropped the commit and added `TypeStorageBucket = 18`

tomponline replied:
thanks!

tomponline commented on lxd/db/cluster/entities.go:
```diff
@@ -81,3 +85,49 @@ func init() {
+	// We need to space separate the path because fmt.Sscanf uses this as a delimiter.
+	spaceSeparatedURLPath := strings.Replace(u.Path, "/", " / ", -1)
+	for entityType, entityURI := range EntityURIs {
+		entityPath, _, _ := strings.Cut(entityURI, "?")
```
Can we use url.Parse here for consistency in extracting the path?

markylaing replied:
This gives a `invalid URL escape "%s" error unfortunately.

Perhaps we should be using regexes here instead.

There are now quite a few places where we need to parse urls and or mux vars into something meaningful outside of the handlers themselves. My initial plan was to utilise the fact that the mux router already does this but it meant parameterising the whole API (naming all routes and having consts for mux vars) but this led to a lot of potentially unnecessary changes and also didn't solve the case for parsing a url contained in an operation resource. In the end, this method seemed a reasonably happy middle ground.

tomponline replied:
Yes this seems fine. Using regexes would be more complicated in my view.

tomponline commented on lxd/db/cluster/entities.go:
```diff
@@ -81,3 +85,49 @@ func init() {
+				pathArgs[i] = *pathArgPtr
+			}
+
+			return entityType, u.Query().Get("project"), pathArgs, nil
```
Should this return "default" for project if not present in URL query string?

markylaing replied:
I've set the project to default if it is not present. It should be unused if the returned entity type isn't project specific. Additionally I've set the project to the project path parameter if the entity type is `TypeProject` in order to avoid confusion.
I've added a comment to explain the above also.

markylaing commented on lxd/db/cluster/entities.go:
```diff
@@ -81,3 +85,49 @@ func init() {
+
+		// Make an []any for the number of expected path arguments and set each value in the slice to a *string.
+		nPathArgs := strings.Count(spaceSeparatedEntityPath, "%s")
+		pathArgsAny := make([]any, nPathArgs)
```
Done.

gabrielmougard commented on lxd/auth/driver_openfga.go:
```diff
@@ -0,0 +1,1079 @@
+	client *client.OpenFgaClient
+}
+
+func (f *fga) configure(opts Opts) error {
```
The type assertion `val.(string)` could potentially panic if the underlying type of val is not a string. It might be safer to use a type assertion with a check, like:
```go
strVal, ok := val.(string)
if !ok {
    return fmt.Errorf("Expected a string for configuration key, got: %T", val)
}
f.apiToken = strVal
```

Also, it might be useful to trim spaces from strings such as the API token, URL, etc., using strings.TrimSpace to avoid potential configuration errors caused by accidental spaces.

gabrielmougard commented on lxd/auth/driver_openfga.go:
```diff
@@ -0,0 +1,1079 @@
+		return fmt.Errorf("Failed to compare OpenFGA model type definitions: %w", err)
+	}
+
+	if string(existingTypeDefinitions) != string(builtinTypeDefinitions) {
```
Is there a risk of false negatives if the field order or whitespace varies but the logical content is the same ?

markylaing replied:
The field order appears to be tied to the model. Both are of the same type and marshaled without indentation so whitespace should be fine. 

I did this because the logic for comparing the models was becoming quite cumbersome but you're right it isn't ideal. I will try again.

gabrielmougard commented on lxd/auth/driver_openfga.go:
```diff
@@ -0,0 +1,1079 @@
+	}
+
+	if readModelResponse.AuthorizationModel.SchemaVersion != builtinAuthorizationModel.SchemaVersion {
+		return fmt.Errorf("Existing OpenFGA model has different schema version")
```
Could we include the actual vs expectedd versions in the error message ?

gabrielmougard commented on lxd/auth/driver_openfga.go:
```diff
@@ -0,0 +1,1079 @@
+	return nil
+}
+
+func (f *fga) projectObjects(ctx context.Context, projectName string) ([]string, error) {
```
Can we refactor using a loop ? For example:

```golang
func (f *fga) projectObjects(ctx context.Context, projectName string) ([]string, error) {
    options := client.ClientListObjectsOptions{
        AuthorizationModelId: openfga.PtrString(f.authModelID),
    }

    objectTypes := []ObjectType{
        ObjectTypeInstance,
        ObjectTypeImage,
        ObjectTypeImageAlias,
        ObjectTypeNetwork,
        ObjectTypeNetworkACL,
        ObjectTypeNetworkZone,
        ObjectTypeProfile,
        ObjectTypeStorageVolume,
        ObjectTypeStorageBucket,
    }

    var allObjects []string
    userObjectString := ObjectProject(projectName).String()

    for _, objectType := range objectTypes {
        resp, err := f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
            User:     userObjectString,
            Relation: relationProject,
            Type:     string(objectType),
        }).Execute()

        if err != nil {
            return nil, err
        }

        allObjects = append(allObjects, resp.GetObjects()...)
    }

    return allObjects, nil
}
```

gabrielmougard commented on lxd/auth/driver_openfga.go:
```diff
@@ -0,0 +1,1079 @@
+	return allObjects, nil
+}
+
+func (f *fga) syncResources(ctx context.Context, resources Resources) error {
```
Same thing here: could we use loops and common pattern and abstract repetitive tasks into helper functions where possible ? For example:

```golang
func (f *fga) syncResources(ctx context.Context, resources Resources) error {
	var writes, deletions []client.ClientTupleKey

	options := client.ClientListObjectsOptions{AuthorizationModelId: openfga.PtrString(f.authModelID)}
	serverUserStr := ObjectServer().String()

	// Helper function to sync resources
	sync := func(resourceType ObjectType, currentResources, existingResources []Object) error {
		for _, currResource := range currentResources {
			if !shared.ValueInSlice(currResource.String(), existingResources) {
				writes = append(writes, client.ClientTupleKey{
					User:     serverUserStr,
					Relation: relationServer,
					Object:   currResource.String(),
				})
			}
		}

		for _, existingResource := range existingResources {
			resourceObject, err := ObjectFromString(existingResource)
			if err != nil {
				return err
			}

			if !shared.ValueInSlice(resourceObject, currentResources) {
				deletions = append(deletions, client.ClientTupleKey{
					User:     serverUserStr,
					Relation: relationServer,
					Object:   existingResource,
				})
			}
		}

		return nil
	}

	resourceMappings := []struct {
		Type       ObjectType
		Resources  []Object
	}{
		{ObjectTypeCertificate, resources.CertificateObjects},
		{ObjectTypeStoragePool, resources.StoragePoolObjects},
		{ObjectTypeProject, resources.ProjectObjects},
	}

	for _, mapping := range resourceMappings {
		resp, err := f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
			User:     serverUserStr,
			Relation: relationServer,
			Type:     string(mapping.Type),
		}).Execute()
		if err != nil {
			return err
		}

		err = sync(mapping.Type, mapping.Resources, resp.GetObjects())
		if err != nil {
			return err
		}
	}

	// Special handling for project resources
	var existingProjectResources []string
	for _, existingProject := range existingProjects {  // Assuming `existingProjects` is fetched somewhere before this line
		projectObject, err := ObjectFromString(existingProject)
		if err != nil {
			return err
		}

		projectResources, err := f.projectObjects(ctx, projectObject.Project())
		if err != nil {
			return err
		}

		existingProjectResources = append(existingProjectResources, projectResources...)
	}

	currentProjectResources := append(resources.ImageObjects, resources.ImageAliasObjects...)
	currentProjectResources = append(currentProjectResources, resources.InstanceObjects...)
	// ... keep appending the other resources
	err := sync(ObjectTypeProject, currentProjectResources, existingProjectResources)
	if err != nil {
		return err
	}

	return f.updateTuples(ctx, writes, deletions)
}
```

markylaing replied:
This is really good feedback, thank you :) I'll work on refining the drivers to remove any repeated logic.

gabrielmougard reviewed (approved):
A couple of refactoring ideas and small details but good overall :+1:

tomponline commented:
This can be rebased now @markylaing

markylaing commented:
> This can be rebased now @markylaing

Yep I'm on it. Just addressing some of @gabrielmougard's comments before testing/pushing.

markylaing commented:
I've rebased and addressed the comments that @gabrielmougard made. 

After addressing those I have added a `TLS` authorization driver to the `OpenFGA` driver. This is used when a user authenticates with TLS. 

After that I fixed the tests a bit, because they were using TLS authentication. The OpenFGA tests now use Candid auth. We can add OIDC later and change it to that so that we test the full range.

@ru-fu @tomponline do you think the documentation should be added to this PR or merged in a follow-up PR?

Let me know what you think, thanks :)

ru-fu commented:
> @ru-fu @tomponline do you think the documentation should be added to this PR or merged in a follow-up PR?

Generally in the same PR, but I'll be OoO until Monday and don't know if I can then get to it right away. So a separate PR might be smarter in this case.

simondeziel commented on test/suites/openfga.sh:
```diff
@@ -0,0 +1,282 @@
+
+user_is_instance_user() {
+  # shellcheck disable=SC3043
+  local instance_name
```
Considering that `local` does nothing and we are not consistently using it, my personal preference would be to stop trying and simply not bother :)

simondeziel commented on test/suites/openfga.sh:
```diff
@@ -0,0 +1,282 @@
+  # Check we can still interact with the instance.
+  LXC_LOCAL='' lxc_remote exec "candid-openfga:${instance_name}" -- echo "bar"
+  touch "${TEST_DIR}/tmp"
+  lxc_remote file push "${TEST_DIR}/tmp" "candid-openfga:${instance_name}/root/tmpfile.txt"
```
I think it'd be safer to remove that `/root/tmpfile.txt`. If you do the removal with a `lxc exec ... -- rm /root/tmpfile.txt` you can drop the earlier `exec ... -- echo "bar"`  test.

simondeziel commented on test/suites/openfga.sh:
```diff
@@ -0,0 +1,282 @@
+  mv "${LXD_CONF}/client.crt.bak" "${LXD_CONF}/client.crt"
+  mv "${LXD_CONF}/client.key.bak" "${LXD_CONF}/client.key"
+  lxc_remote config trust remove "candid-openfga:${test_cert_fingerprint}"
+  lxc remote rm test-remote
```
s/rm/delete/, please.

simondeziel reviewed (commented):
Very nice and thorough tests!

markylaing commented on test/includes/certificates.sh:
```diff
@@ -0,0 +1,4 @@
+cert_fingerprint() {
+  openssl x509 -noout -fingerprint -sha256 -in "${1}" |
+    sed 's/.*Fingerprint=//; s/://g; s/\([A-F0-9]\{64\}\).*/\1/' | tr '[:upper:]' '[:lower:]'
```
Nice! I didn't know this :smile:

markylaing commented on test/suites/openfga.sh:
```diff
@@ -0,0 +1,282 @@
+  user_is_project_operator
+
+  # Change permission to "user" for instance "foo"
+  lxc launch testimage foo
```
The `user_is_instance_user` test below needs to check `exec` and `file push` so the instance needs to be running

tomponline replied:
For containers you acruslky don't need the instance running to do that btw

markylaing replied:
I'm confused. Surely the instance needs to be running to run `lxc exec`? LXD errors with `Error: Instance is not running`.

tomponline replied:
File push doesn't need instance running

markylaing replied:
Ah ok. The test below needs to check both and the user doesn't have permission to start the instance within the test so it does need to be `launch` here. I've changed it to `init` where appropriate.

simondeziel replied:
Sorry, I hadn't notice the `lxc exec` bit.

markylaing commented on test/suites/openfga.sh:
```diff
@@ -0,0 +1,282 @@
+
+  echo "==> Checking permissions for instance user..."
+  user_is_instance_user foo
+  lxc delete foo --force
```
I'll move it down a little but the instance is required only for the instance user tests. I'll rename it `user-foo` and add more comments.

tomponline replied:
My preference would be to remove it as soon as its no longer needed.

markylaing replied:
This didn't work because the subsequent tests assume a clean default project. I've added comments to explain.

markylaing commented on test/suites/openfga.sh:
```diff
@@ -0,0 +1,282 @@
+
+  fingerprint="$(lxc config trust list -f csv | cut -d, -f4)"
+
+  # But cannot edit non-certificate fields.
```
Do you mean although the user is not an admin, they can still change their own certificate? If so, doing that via yaml will be pretty gnarly... I could at least double check that they are able to revoke their own certificate though.

markylaing replied:
Ah of course, this no longer makes sense because we're using candid authentication. The certificate we're deleting isn't "ours" in the sense that it's what is used to authenticate when we use the `localhost` remote, not the `candid-openfga` remote.

tomponline commented:
Needs a rebase to solve conflicts

tomponline commented on .github/workflows/tests.yml:
```diff
@@ -249,6 +249,12 @@ jobs:
           curl -sSfL https://dl.min.io/server/minio/release/linux-amd64/minio --output "$(go env GOPATH)/bin/minio"
           chmod +x "$(go env GOPATH)/bin/minio"
+          
+          curl -L https://github.com/openfga/openfga/releases/download/v1.3.1/openfga_1.3.1_linux_amd64.tar.gz -o openfga.tar.gz
```
Is there a link for latest version we can use rather than having to keep remembering to update this?

simondeziel replied:
I checked that but since they put the version number in the asset themselves, it didn't work in my tests.

tomponline commented on .github/workflows/tests.yml:
```diff
@@ -249,6 +249,12 @@ jobs:
+          curl -L https://github.com/openfga/openfga/releases/download/v1.3.1/openfga_1.3.1_linux_amd64.tar.gz -o openfga.tar.gz
+          tar -xzf openfga.tar.gz -C "$(go env GOPATH)/bin/"
+          
+          curl -L https://github.com/openfga/cli/releases/download/v0.1.0/fga_0.1.0_linux_amd64.tar.gz -o fga.tar.gz
```
Is there a link for latest version we can use rather than having to keep remembering to update this?

markylaing replied:
Yep I've figured that out just now, should be able to push shortly.

markylaing replied:
I've changed these to:
```
# Download latest release of openfga server.
curl -s https://api.github.com/repos/openfga/openfga/releases/latest | jq -r '.assets | .[] | .browser_download_url | select(. | test("_linux_amd64.tar.gz$"))' | xargs -I {} curl -L {} -o openfga.tar.gz
tar -xzf openfga.tar.gz -C "$(go env GOPATH)/bin/"

# Download latest release of openfga cli.
curl -s https://api.github.com/repos/openfga/cli/releases/latest | jq -r '.assets | .[] | .browser_download_url | select(. | test("_linux_amd64.tar.gz$"))' | xargs -I {} curl -L {} -o fga.tar.gz
tar -xzf fga.tar.gz -C "$(go env GOPATH)/bin/"
```
It's not particularly readable but it gets the latest release via the github API and downloads it in a one-liner.

markylaing commented:
I have added reference and explanation documentation. I think this will also need a how-to but I need to figure out how to easily set up an OIDC server in order to do that (keycloak requires a lot of clicking around and openjdk-17).

tomponline commented on lxd/daemon.go:
```diff
@@ -1802,6 +1803,246 @@ func (d *Daemon) Stop(ctx context.Context, sig os.Signal) error {
 	return err
 }
 
+// Setup OpenFGA.
```
A docblock comment explaining what this function does would be nice.

tomponline commented on lxd/daemon.go:
```diff
@@ -1336,6 +1337,20 @@ func (d *Daemon) init() error {
+		if openFGAAuthorizationModelID == "" {
+			// We should never be missing the model ID at start up if we have other connection details (this means
+			// something went wrong the last time we tried to set it up).
+			logger.Warn("OpenFGA authorization driver is misconfigured, skipping...")
```
So this will fallback to TLS only mode or only allow unix socket operation?

tomponline replied:
I think either both should be warnings or both should be errors, as they both result in the authorizer not working right?

tomponline commented on lxd/daemon.go:
```diff
@@ -1336,6 +1337,20 @@ func (d *Daemon) init() error {
+		} else {
+			err = d.setupOpenFGA(openfgaAPIURL, openfgaAPIToken, openfgaStoreID, openFGAAuthorizationModelID)
+			if err != nil {
+				logger.Error("Failed to configure OpenFGA. Reverting to default TLS authorization", logger.Ctx{"error": err})
```
Same question as above

tomponline commented on lxd/api_1.0.go:
```diff
@@ -808,6 +813,50 @@ func doApi10Update(d *Daemon, r *http.Request, req api.ServerPut, patch bool) re
 	return response.EmptySyncResponse
 }
 
+func doApi10PreNotifyTriggers(d *Daemon, clusterChanged map[string]string, newClusterConfig *clusterConfig.Config) error {
```
missing docblock explaining what this is and what it does

tomponline commented on lxd/api_1.0.go:
```diff
@@ -763,6 +763,11 @@ func doApi10Update(d *Daemon, r *http.Request, req api.ServerPut, patch bool) re
 		}
 	})
 
+	err = doApi10PreNotifyTriggers(d, clusterChanged, newClusterConfig)
```
Do we need to support a revert here if something fails later on?

tomponline reviewed (approved):
Thanks, this is great work.
Please can you address/discuss the changes via your existing github card for further minor work.

But I'll merge this now as I can't see any major issues.

Thanks for your thoroughness.
